pkg/obfuscate/sql.go

```diff
@@ -220,90 +313,35 @@ func (o *Obfuscator) obfuscateSQLString(in string, opts *SQLConfig) (*Obfuscated
 	return out, err
 }
 
-// tableFinderFilter is a filter which attempts to identify the table name as it goes through each
-// token in a query.
-type tableFinderFilter struct {
-	storeTableNames bool
-	// seen keeps track of unique table names encountered by the filter.
-	seen map[string]struct{}
-	// csv specifies a comma-separated list of tables
-	csv strings.Builder
-}
-
-// Filter implements tokenFilter.
-func (f *tableFinderFilter) Filter(token, lastToken TokenKind, buffer []byte) (TokenKind, []byte, error) {
-	switch lastToken {
-	case From, Join:
-		// SELECT ... FROM [tableName]
-		// DELETE FROM [tableName]
-		// ... JOIN [tableName]
-		if r, _ := utf8.DecodeRune(buffer); !unicode.IsLetter(r) {
-			// first character in buffer is not a letter; we might have a nested
-			// query like SELECT * FROM (SELECT ...)
-			break
-		}
-		fallthrough
-	case Update, Into:
-		// UPDATE [tableName]
-		// INSERT INTO [tableName]
-		if f.storeTableNames {
-			f.storeName(string(buffer))
-		}
-		return TableName, buffer, nil
-	}
-	return token, buffer, nil
-}
-
-// storeName marks the given table name as seen in the internal storage.
-func (f *tableFinderFilter) storeName(name string) {
-	if _, ok := f.seen[name]; ok {
-		return
-	}
-	if f.seen == nil {
-		f.seen = make(map[string]struct{}, 1)
-	}
-	f.seen[name] = struct{}{}
-	if f.csv.Len() > 0 {
-		f.csv.WriteByte(',')
-	}
-	f.csv.WriteString(name)
-}
-
-// CSV returns a comma-separated list of the tables seen by the filter.
-func (f *tableFinderFilter) CSV() string { return f.csv.String() }
-
-// Reset implements tokenFilter.
-func (f *tableFinderFilter) Reset() {
-	for k := range f.seen {
-		delete(f.seen, k)
-	}
-	f.csv.Reset()
-}
-
 // ObfuscatedQuery specifies information about an obfuscated SQL query.
 type ObfuscatedQuery struct {
-	Query     string // the obfuscated SQL query
-	TablesCSV string // comma-separated list of tables that the query addresses
+	Query    string      `json:"query"`    // the obfuscated SQL query
+	Metadata SQLMetadata `json:"metadata"` // metadata extracted from the SQL query
 }
 
 // Cost returns the number of bytes needed to store all the fields
 // of this ObfuscatedQuery.
 func (oq *ObfuscatedQuery) Cost() int64 {
-	return int64(len(oq.Query) + len(oq.TablesCSV))
+	return int64(len(oq.Query) + len(oq.Metadata.TablesCSV))
```

Comment:
This becomes invalid now. It should contain the length of all the strings now (including comments & commands). This method is called when performance is critical and a cache is used, so iterating here over two slices will not be a good idea. Instead, it would be nice to store the cost inside `SQLMetadata` and increase it dynamically when comments & commands are added. To do this nicely, I'd recommend removing the fields from `metadataFinderFilter` and having just one `SQLMetadata` field which stores all the information, having methods to add comments & commands, which in turn automatically increase the cost. Within that context, it will be better to call it `Size` instead of `Cost`.

Reply:
How about computing the cost on the fly? I mean simply leaving things as they are, and adding a new field into the filter called `resultBytes` or something, and setting that as `SQLMetadata.Cost` at the end.